### ISSUE TITLE
Explicit `iter` call required for `__getitem__` with overloads

### ISSUE THREAD

**Issue by shiftinv** [microsoft/pyright#3702]
**Describe the bug**
When a class has overloads for `__getitem__`, iterating over an instance of this class requires using `for x in iter(inst):` instead of `for x in inst:`.

**To Reproduce**
```py
import typing as t

class Clazz:
    @t.overload
    def __getitem__(self, i: int) -> str:
        ...

    @t.overload
    def __getitem__(self, i: slice) -> t.List[str]:
        ...

    def __getitem__(self, i: t.Union[int, slice]) -> t.Union[str, t.List[str]]:
        ...

c = Clazz()

for x in iter(c): ...  # (no errors, type of `x` inferred correctly)
for x in c: ...  # "Clazz" is not iterable - "__iter__" method not defined
```

**Expected behavior**
Iterating over an instance of a class that defines `__getitem__` with overloads should work the same, no matter whether `iter` and/or overloads are used.
Without the overloads or when using `iter`, no errors are shown, as expected.

**VS Code extension or command-line**
cli, 1.1.260

**Comment by erictraut** [comment]
This is included in pyright 1.1.261, which I just published. It will also be included in an upcoming release of pylance.
